src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -2072,6 +2132,30 @@ static void start_batch_in_call_combiner(void* arg, grpc_error* ignored) {
   grpc_subchannel_call_process_op(subchannel_call, batch);
 }
 
+// Adds a closure to closures that will execute batch in the call combiner.
+static void add_closure_for_subchannel_batch(
+    call_data* calld, grpc_transport_stream_op_batch* batch,
+    closure_to_execute* closures, size_t* num_closures) {
+  batch->handler_private.extra_arg = calld->subchannel_call;
+  GRPC_CLOSURE_INIT(&batch->handler_private.closure,
+                    start_batch_in_call_combiner, batch,
+                    grpc_schedule_on_exec_ctx);
+  closure_to_execute* closure = &closures[(*num_closures)++];
+  closure->closure = &batch->handler_private.closure;
+  closure->error = GRPC_ERROR_NONE;
+  // If the tracer is enabled, we log a more detailed message, which
```

Comment:
The flagz mechanism is a horrible idea, one that SRE strongly discourages the use of.  Allowing people to change flags at run-time without code review or any audit trail means that you can't have any idea what state the server is actually in, and that state can be completely lost any time the server restarts.  It's also not a reliable mechanism, because there is no way to tell whether a given flag is consulted at run-time or only once at startup.

I think we should not support changing tracer state as the program runs.  Doing so would make it subject to this exact same set of problems.

With regard to this specific code, I don't see a better way to do this.  If the tracer is enabled, I do want detailed messages that help with debugging, but I don't want the overhead of dynamic allocation if the messages are not going to be seen.  I'd welcome suggestions on a better way to do this.

Reply:
By doing extra memory allocation, tracers can affect the program's performance.  I don't think this is really any different from the flagz issue; performance should be considered part of the program's "fundamental state".  The principle here is that that state should not be alterable from outside of the program's normal execution flow.

In any case, I've changed the code to indicate whether the string needs to be freed for each individual entry, so it should be safe either way.